### DISCUSSION TITLE
What is the story behind ACTIONS_RUNTIME_URL, ACTIONS_RUNTIME_TOKEN, GITHUB_RUN_ID

### DISCUSSION THREAD

**Discussion by tapika** [cake-build/cake#3783] [General]
I have tried to perform artifact upload from github actions:

`context.BuildSystem().GitHubActions.Commands.UploadArtifact`

And apparently I need to have following environment variables to be defined:

```
1. ACTIONS_RUNTIME_TOKEN
2. GITHUB_RUN_ID
3. ACTIONS_RUNTIME_URL
```

With 1 - I think it's more or less obvious - I could remap `GITHUB_TOKEN` to be `ACTIONS_RUNTIME_TOKEN` - don't understand why, but in theory this could be done.

```
env:
  ACTIONS_RUNTIME_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

With 2 - it's also more or less ok - `RUN_ID` is just some unique run number. If I want to test on my own PC, I could just use `1` maybe or just not to test on my PC.

But with 3 it's completely unclear what to use as URL. `https://github.com/<owner>/<repo>` ? 

Also based on some of documentation - I suspect that 1 & 2 should be defined by github itself, and they are some of undocumented environment variables. I suspect those are only exposed to github action runner, not to batch / ps1 script child process which gets executed by github actions.

Do you have any documentation which could give some highlight on how `context.BuildSystem().GitHubActions.Commands.UploadArtifact`
should work ?

**Comment by devlead** [comment]
If you use the Cake GitHub Action these will be populated automatically

https://github.com/cake-build/cake-action

Example:
```yaml
steps:
  - name: Run the Cake script
    uses: cake-build/cake-action@v1
```

You can see example usage of the GitHub provider commands in our integration tests 

https://github.com/cake-build/cake/blob/a50935365b1206f4e8af0e613b6c82e4b53c1445/tests/integration/Cake.Common/Build/GitHubActions/GitHubActionsProvider.cake#L31-L49

**Reply by tapika** [reply]
What if I'm using Cake Frosten, and launch cake as a batch or powershell script ?

Is there any way I can pass those environment variables from Github actions application to child process ?

**Reply by devlead** [reply]
GitHub Script/Bash/PowerShell steps, unfortunately, won't pass these variables to scripts/child processes.

Cake Frosting support for Cake/Action is something we're investigating/thinking about, meanwhile, you could use Cake.Tool to bootstrap Frosting from a small Cake script which you call from Cake/Action.

Such a script could look something like
```cake
DotNetRun(
    "./build/Build.csproj",
    new ProcessArgumentBuilder()
        .AppendSwitch("--target", "=", Argument("target", "Default")));
```
and YAML to call it something like
```YAML
      - name: Run Cake script
        uses: cake-build/cake-action@v1
        with:
          target: GitHubActions
```

Complete example repo of how that can be done here:
https://github.com/devlead/CakeFrostingAction

**Reply by tapika** [reply]
Hmm... I've managed to reconfigure cake to provide environment variables to Cake frosten.

Works like this - add following step:

```
    - name: Github actions environment to local environment
      uses: cake-build/cake-action@v1
      with:
        script-path: routeenvvar.cake
```

And add following script:

routeenvvar.cake:
```
// Sets environment variables available from Github action process to child processes
var file = System.Environment.GetEnvironmentVariable("GITHUB_ENV");
System.IO.File.AppendAllText(file, $"ACTIONS_RUNTIME_URL={System.Environment.GetEnvironmentVariable("ACTIONS_RUNTIME_URL")}\n");
System.IO.File.AppendAllText(file, $"ACTIONS_RUNTIME_TOKEN={System.Environment.GetEnvironmentVariable("ACTIONS_RUNTIME_TOKEN")}\n");
System.Environment.Exit(0);
```

Looks like working, what I have initially tried on. (In https://github.com/tapika/swupd , develop branch).

**Reply by devlead** [reply]
If you launch from a cake script like I did above it'll be available to all child processes.

**Reply by devlead** [reply]
Also setting github actions environment variables using Cake

```cake 
GitHubActions.Commands.SetEnvironmentVariable(key, value);
```

**Reply by tapika** [reply]
Thanks, I will probably try other approaches later on. Would be nice to find out fastest from them.
